File: iOS/LeetCodeSwift/LeetCodeSwift/Leetcode/Linked_List__Cycle.swift
//
//  Linked_List__Cycle.swift
//  LeetCodeSwift
//
//  Created by tree on 2018/3/20.
//  Copyright © 2018年 treee. All rights reserved.
//

import Foundation

extension Solution {
    
    /**
     我们设置快慢两个游标，两个游标速度不同
     如果是链表中含有环的话，那么两个指针肯定会最终相等
     */
    func has_cycle(_ head: ListNode) -> Bool {
        var slow: ListNode? = head
        var fast: ListNode? = head
        while fast != nil && fast?.next != nil {
            slow = slow?.next
            fast = fast?.next?.next
            if slow! == fast! { return true }
        }
        return false
    }
}

File: python/LeetCode/Linked_List_Cycle.py
from Structs import ListNode

class Solution:
    """
    判断一个链表结构是否有环
     思路： 设置两个指针，一快一慢，
     如果这个链表有环的话那么指针一定会走到同一处

     如果遍历到了最后都没有遇到，说明结构无环

    """
    def hasCycle(self, head) -> bool:
        if head is None: return False
        slow, fast = head, head
        while fast is not None \
                and fast.next is not None:
            slow = slow.next
            fast = fast.next.next
            if slow is fast: return True
        return False